toponetx/utils/normalization.py

```diff
@@ -0,0 +1,393 @@
+
+    Notes
+    -----
+    This normalization is based on Kipf's formulation, which computes the row-sums,
+    constructs a diagonal matrix D^(-0.5), and applies the normalization as D^(-0.5) * A_opt * D^(-0.5).
+    """
+    if add_identity:
+        size = A_opt.shape[0]
+        eye = diags(size * [identity_multiplier])
+        A_opt = np.abs(A_opt) + eye
+    else:
+        A_opt = np.abs(A_opt)
+    rowsum = np.array(A_opt.sum(1))
+    r_inv_sqrt = np.power(rowsum, -0.5).flatten()
+    r_inv_sqrt[np.isinf(r_inv_sqrt)] = 0.0
+    r_mat_inv_sqrt = diags(r_inv_sqrt)
+    A_opt_normalized = A_opt.dot(r_mat_inv_sqrt).transpose().dot(r_mat_inv_sqrt)
+
+    return A_opt_normalized
+
+
+def compute_xu_asymmetric_normalized_matrix(
+    B: np.ndarray | csr_matrix, is_sparse: bool = True
+) -> np.ndarray | csr_matrix:
+    """Compute Xu's normalized asymmetric matrix.
+
+    Typically used to normalize boundary operators.
+
+    Parameters
+    ----------
+    B : np.ndarray or csr_matrix
+        The asymmetric matrix.
+    is_sparse : bool, default=True
+        If True, treat B as a sparse matrix.
+
+    Returns
+    -------
+    np.ndarray or csr_matrix
+        The normalized asymmetric matrix.
+
+    Notes
+    -----
+    This normalization is based on Xu's formulation, which computes the diagonal matrix D^(-1)
+    and multiplies it with the asymmetric matrix B.
+    """
+    if is_sparse:
+        rowsum = np.array(np.abs(B).sum(1))
+        r_inv = np.power(rowsum, -1).flatten()
+        r_inv[np.isinf(r_inv)] = 0.0
+        r_mat_inv = diags(r_inv)
+        normalized_matrix = r_mat_inv.dot(B)
+        return normalized_matrix
+    else:
+        Di = np.sum(np.abs(B), axis=1)
+        Di2 = np.diag(np.power(Di, -1))
+        Di2[np.isinf(Di2)] = 0.0
+        B = np.array(B)
+        normalized_matrix = Di2.dot(B)
+        return normalized_matrix
+
+
+def compute_bunch_normalized_matrices(
+    B1: np.ndarray | csr_matrix, B2: np.ndarray | csr_matrix
+) -> Tuple[
+    np.ndarray | csr_matrix,
+    np.ndarray | csr_matrix,
+    np.ndarray | csr_matrix,
+    np.ndarray | csr_matrix,
+]:
+    """Get Bunch normalization.
+
+    Parameters
+    ----------
+    B1 : np.ndarray or csr_matrix
+        The boundary B1: C1->C0 of a complex.
+    B2 : np.ndarray or csr_matrix
+        The boundary B2: C2->C1 of a complex.
+
+    Returns
+    -------
+    B1 : np.ndarray or csr_matrix
+        Normalized B1: C1->C0.
+    B1T : np.ndarray or csr_matrix
+        Normalized B1T: C0->C1.
+    B2 : np.ndarray or csr_matrix
+        Normalized B2: C2->C1.
+    B2T : np.ndarray or csr_matrix
+        Normalized B2T: C1->C2.
+
+    References
+    ----------
+    [1] Schaub, M. T., Benson, A. R., Horn, P., Lippner, G., & Jadbabaie, A.
+        "Random walks on simplicial complexes and the normalized hodge 1-laplacian."
+
+    [2] Bunch, E., You, Q., Fung, G., & Singh, V.
+        "Simplicial 2-Complex Convolutional Neural Networks."
+    """
```

Comment:
The references section is still not in the required format. See https://github.com/pyt-team/TopoModelX/blob/main/topomodelx/nn/simplicial/dist2cycle_layer.py for an example or here https://bashtage.github.io/sphinx-material/rst-cheatsheet/rst-cheatsheet.html